src/js/legend.js

```diff
@@ -0,0 +1,227 @@
+            // Or, a d3 color scale that's generated based off an array of colors
+            colors: null,
+            sort: true, // Whether or not to sort the data by values when creating the legend
+            showLegendHeadings: true // Whether or not to display column headings in the legend
+        },
+        modelRelay: [{
+            target: "dataSetWithColors",
+            singleTransform: {
+                type: "fluid.transforms.free",
+                args: {
+                    "dataSet": "{that}.model.dataSet",
+                    "colors": "{that}.options.legendOptions.colors"
+                },
+                func: "gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors"
+            }
+        }],
+        styles: {
+            legend: "gpii-ca-pieChart-legend",
+            table: "gpii-ca-pieChart-legend-table",
+            row: "gpii-ca-pieChart-legend-table-row",
+            colorCell: "gpii-ca-pieChart-legend-color-cell",
+            labelCell: "gpii-ca-pieChart-legend-label-cell",
+            valueCell: "gpii-ca-pieChart-legend-value-cell"
+        },
+        selectors: {
+            legend: ".gpiic-ca-pieChart-legend",
+            table: ".gpii-ca-pieChart-legend-table",
+            row: ".gpii-ca-pieChart-legend-table-row",
+            colorCell: ".gpii-ca-pieChart-legend-color-cell",
+            labelCell: ".gpii-ca-pieChart-legend-label-cell",
+            valueCell: ".gpii-ca-pieChart-legend-value-cell"
+        },
+        events: {
+            onLegendCreated: null  // Fire when the legend is created. Ready to register D3 DOM event listeners
+        },
+        listeners: {
+            "onCreate.create": {
+                funcName: "gpii.chartAuthoring.pieChart.legend.create",
+                args: ["{that}"]
+            }
+        },
+        modelListeners: {
+            dataSet: {
+                funcName: "{that}.draw",
+                excludeSource: "init"
+            }
+        },
+        invokers: {
+            draw: {
+                funcName: "gpii.chartAuthoring.pieChart.legend.draw",
+                args: ["{that}"]
+            },
+            sort: {
+                funcName: "gpii.chartAuthoring.pieChart.legend.sort",
+                args: ["{arguments}.0", "{arguments}.1"]
+            },
+            getColorCellStyle: {
+                funcName: "gpii.chartAuthoring.pieChart.legend.getColorCellStyle",
+                args: ["{arguments}.0"]
+            }
+        }
+    });
+
+    // Takes the dataSet array and the color array, and returns a consolidated object array to ease sorting and other operations while keeping colors "correct"
+    gpii.chartAuthoring.pieChart.legend.consolidateDataAndColors = function (model) {
+        var dataSet = model.dataSet;
+        var colors = (typeof(model.colors) === "function") ? model.colors : gpii.d3.getColorScale(model.colors);
+
+        var consolidated = [];
+        fluid.each(dataSet, function(item, i) {
```

Comment:
Another case where fluid.transform is probably more appropriate. Also you could use `fluid.copy` to make a copy of the model.

```javascript
return fluid.transform(dataSet, function (data, idx) {
    var consolidated = fluid.copy(data);
    consolidated["color"] = colors(idx);
    return consolidated;
});